### PR TITLE
feat: Mempool iteration refactoring, with better scaling in the size of the mempool

### PR DESCRIPTION
**Pull request by gregorycoppola**
Note: Does NOT use RAM to store the mempool.

### Description
This algorithm implements the mempool walk to consider candidates in time `O(N)` in the size of the mempool. In addition, it only does `O(M)` deserializations, where `M` is the number of transactions that 1) meet the criteria, and 2) can be processed in the available time.

This replaces an algorithm which in these terms would be `O(MN^2)`. We want the mempool size (i.e., `N`) to be able to get quite large, and the baseline system can degrade at around 10k-100k transactions.

### Algorithm
The algorithm proceeds as:
* select a very "thin" (minimal) representation of each transaction in the entire mempool
* sort this list by fee rate
* for each `txid` in the sorted order:
  * check the nonce appropriateness, and `continue` if not appropriate
  * fetch and deserialize the associated `tx`
  * process this `tx`

### Evaluation
This algorithm was evaluated and developed in https://www.notion.so/hiropbc/Productionization-of-In-Memory-Mempool-c0599678903341d1bd70a728b0e1ab85

However, I am still testing to see that this PR version will get the same results as the previous tests.

### Applicable issues
- towards #130

### Additional info (benefits, drawbacks, caveats)
Caveats:
* currently does not support sponsor nonces (I wasn't sure how to handle the None case reading from DB)

### Checklist
- [ ] Test coverage for new or modified code paths
- [ ] Changelog is updated
- [ ] New integration test(s) added to `bitcoin-tests.yml`

### CODE REVIEW

**Comment by gregorycoppola** [comment]
There are still some small test breaks, but I think one could start reviewing this PR now for code, algo, etc.

File: src/core/mempool.rs [hunk 1]
```diff
@@ -236,6 +237,17 @@ pub struct MemPoolTxInfo {
     pub metadata: MemPoolTxMetadata,
 }
 
+/// This class is a minimal version of `MemPoolTxInfo`. It contains
+/// just enough information to 1) filter by nonce readiness, 2) sort by fee rate.
+#[derive(Debug, PartialEq, Clone)]
+pub struct MemPoolTxMinimalInfo {
```
Comment (obycode) [inline src/core/mempool.rs:243]: Should we just use `MemPoolTxMetadata` here?

**Comment by obycode** [comment]
Are these changes replaced with our work upstream in stacks-blockchain @gregorycoppola?

**Comment by gregorycoppola** [comment]
@obycode this can be replaced with the version checked in to mainnet... would make sense to do that for consistency.

**Pull request closed without merging**